src/Cache/Cache.php

```diff
@@ -1,104 +1,195 @@
-            return $result;
-        }, []);
+        $data = [];
+
+        foreach ($this->fromIterable($keys) as $key) {
+            $data[$key] = $this->get($key, $default);
+        }
+
+        return $data;
+    }
+
+    /** @inheritdoc */
+    public function setMultiple(iterable $values, int|DateInterval|null $ttl = null): bool
+    {
+        $return = [];
+
+        foreach ($this->fromIterable($values) as $key => $value) {
+            $return[] = $this->set($key, $value, $ttl);
+        }
+
+        return $this->checkReturn($return);
+    }
+
+    /** @inheritdoc */
+    public function deleteMultiple(iterable $keys): bool
+    {
+        $return = [];
+
+        foreach ($this->fromIterable($keys) as $key) {
+            $return[] = $this->delete($key);
+        }
+
+        return $this->checkReturn($return);
     }
 
     /**
-     * @param array<array{key:string, value:string}> $values
-     * @param \DateInterval|int|null $ttl
-     * @return bool
-     * @throws CacheException
+     * @throws \InvalidArgumentException
      */
-    public function setMultiple(iterable $values, \DateInterval|int|null $ttl = null): bool
+    protected function checkKey(string $key): string
     {
-        foreach ($values as $key => $value) {
-            $this->set($key, $value, $ttl);
+
+        if (empty($key)) {
+            throw new InvalidArgumentException('cache key is empty');
         }
-        return true;
+
+        return $key;
     }
 
-    public function deleteMultiple(iterable $keys): bool
+    /**  */
+    protected function checkKeyArray(array $keys): array
     {
+
         foreach ($keys as $key) {
-            unset($this->cache_db[$key]);
+            $this->checkKey($key);
         }
+
+        return $keys;
+    }
+
+    /**
+     * @throws \InvalidArgumentException
+     */
+    protected function fromIterable(iterable $data): array
+    {
+
+        if (is_array($data)) {
+            return $data;
+        }
+
+        if ($data instanceof Traversable) {
+            return iterator_to_array($data); // @codeCoverageIgnore
+        }
+
+        throw new InvalidArgumentException('invalid data');
+    }
+
+    /**  */
+    protected function getTTL(DateInterval|int|null $ttl): ?int
+    {
+
+        if ($ttl instanceof DateInterval) {
+            return (new DateTime())->add($ttl)->getTimeStamp() - time();
+        }
+
+        // We treat 0 as a valid value.
+        if (is_int($ttl)) {
+            return $ttl;
+        }
+
+        return null;
+    }
+
+    /**
+     * @param bool[]|int[] $booleans
+     */
+    protected function checkReturn(array $booleans): bool
+    {
+
+        foreach ($booleans as $boolean) {
+            if (!(bool)$boolean) {
+                return false; // @codeCoverageIgnore
+            }
+        }
+
         return true;
     }
 
-    public function has(string $key): bool
+    /**
+     * Get all cache keys. Needed for testing purposes.
+     *
+     * @return array{string}
```

Comment:
this should be `string[]` or `array<string>`. `array{string}` is an array shape with a single key 0 being a string, which is not what this method returns.